### PR TITLE
Add BufferDescriptor API to Material::Builder.

### PR DESCRIPTION
**Pull request by prideout**
@pixelflinger, please take a look since you might want to clean this up after it goes in, I decided to avoid touching `BuilderBase`.

Per discussion with Romain, this is motivated by our upcoming JS bindings, which prefer dealing with buffer descriptors instead of void * style API.

### CODE REVIEW

File: filament/src/Material.cpp [hunk 1]
```diff
@@ -45,10 +45,14 @@ using namespace driver;
...
-    const void* mPayload = nullptr;
-    size_t mSize = 0;
     filaflat::MaterialParser* mMaterialParser = nullptr;
     bool mDefaultMaterial = false;
+
+    // We have to use the heap to stash the buffer descriptor because BuilderBase uses copy
+    // assignment on the details instance, and BufferDescriptor has a deleted copy constructor. For
+    // the same reason, we also need to use shared_ptr rather than unique_ptr.
+    // TODO: consider refactoring BuilderBase to allow descendants that do not support copying.
+    std::shared_ptr<BufferDescriptor> mBufferDescriptor = std::make_shared<BufferDescriptor>();
```
Comment (romainguy) [inline filament/src/Material.cpp]: Eeewwww. @pixelflinger?

File: filament/include/filament/Material.h [hunk 2]
```diff
@@ -87,10 +89,15 @@ class UTILS_PUBLIC Material : public FilamentAPI {
         Builder& operator=(Builder const& rhs) noexcept;
         Builder& operator=(Builder&& rhs) noexcept;
 
-        // This does not copies the content of the RAM, only copy references.
+        // This does not copy the content of the RAM, it only makes a weak reference.
         // The RAM must stay valid until build() is called.
         Builder& package(const void* payload, size_t size);
 
+        // Alternatively, clients can submit a BufferDescriptor instead of void *, which is
+        // auto-released if a callback is present. Note that some language bindings (e.g.,
```
Comment (romainguy) [inline filament/include/filament/Material.h]: I'm not sure I understand. BufferDescriptor does not release the memory, even when a callback is present.

File: filament/src/Material.cpp [hunk 3]
```diff
@@ -64,20 +68,28 @@ BuilderType::Builder& BuilderType::Builder::operator=(BuilderType::Builder const
 BuilderType::Builder& BuilderType::Builder::operator=(BuilderType::Builder&& rhs) noexcept = default;
 
 Material::Builder& Material::Builder::package(const void* payload, size_t size) {
-    mImpl->mPayload = payload;
-    mImpl->mSize = size;
+    mImpl->mBufferDescriptor.reset(new BufferDescriptor(payload, size));
```
Comment (pixelflinger) [inline filament/src/Material.cpp]: I don't want to do a heap allocation in the common case.

File: filament/src/Material.cpp (hunk 1 above)
Comment (pixelflinger) [inline filament/src/Material.cpp]: You could have a copy-ctor on BuilderDetails instead, e.g.:
```
BuilderDetails::BuilderDetails(BuilderDetails const& rhs) 
    : mBufferDescriptor(rhs.data, rhs.size, rhs.callback) {
}
```

You may have to make BufferDescriptor::callback public or add a getCallback().

File: filament/src/Material.cpp [hunk 4]
```diff
@@ -64,20 +68,28 @@ BuilderType::Builder& BuilderType::Builder::operator=(BuilderType::Builder const
...
+    *mImpl->mBufferDescriptor = std::move(buffer);
     return *this;
 }
 
 Material* Material::Builder::build(Engine& engine) {
-    MaterialParser* materialParser = new MaterialParser(
-            upcast(engine).getBackend(), mImpl->mPayload, mImpl->mSize);
+    // By moving the stashed buffer descriptor into a temporary, we are ensuring that its
+    // user-supplied callback is called when execution leaves this scope.
+    BufferDescriptor bd = std::move(*mImpl->mBufferDescriptor);
```
Comment (pixelflinger) [inline filament/src/Material.cpp]: I think this doesn't work if:

- you did a copy of the builder. The shared-pointer will track the buffer correctly, but this line will destroy the buffer right here.
- you called `.build()` more than once

File: samples/vk_hellotriangle.cpp [hunk 5]
```diff
@@ -54,10 +54,12 @@ static const Vertex TRIANGLE_VERTICES[3] = {
 
 static constexpr uint16_t TRIANGLE_INDICES[3] = { 0, 1, 2 };
 
-static constexpr uint8_t BAKED_COLOR_PACKAGE[] = {
+static constexpr uint8_t BAKED_COLOR[] = {
     #include "generated/material/bakedColor.inc"
 };
 
+static auto BAKED_COLOR_PACKAGE = Material::BufferDescriptor(BAKED_COLOR, sizeof(BAKED_COLOR));
```
Comment (pixelflinger) [inline samples/vk_hellotriangle.cpp]: why this change? I don't think we want that. No reason to go through the BufferDescriptor -- which does heap allocation, etc...
